src/frontend/src/optimizer/plan_node/batch_hash_join.rs

```diff
@@ -162,20 +165,13 @@ impl ToDistributedBatch for BatchHashJoin {
     fn to_distributed(&self) -> Result<PlanRef> {
         let right = self.right().to_distributed_with_required(
             &Order::any(),
-            &RequiredDist::shard_by_key(
-                self.right().schema().len(),
-                &self.eq_join_predicate().right_eq_indexes(),
-            ),
+            &RequiredDist::hash_shard(&self.eq_join_predicate().right_eq_indexes()),
```

Comment:
it works, but still not good enough. which might insert unnecessary exchange. May be we can still the require `ShardByKey` and then check the plan's distribution. if the distribution has a `UpstreamHashShard` we can enforce it to hash shard then.

Reply:
```
    let left = left.to_distributed_with_required(shard_by_key(join_key));
    match left.distribution{
      HashShard(s) => right = right.to_distributed_with_required(s);
      UpstreamHashShard(s) => {
           right = right.to_distributed_with_required(shard_by_key(join_key));
           match right.distribution{
               HashShard(s) => left = s.enforce(left);
               UpstreamHashShard(s) => {
                       left = hash_shard(join_key).enforce(left);
                       right = hash_shard(join_key).enforce(right);
               }
               _ => unreachable!()
           }
      }
      _ => unreachable!()
    }
```
just pseudocode

Reply:
I agree, but as mentioned in PR description, I'm not sure whether to do it 🤔

Reply:
The problem is: since an `UpstreamHashShard` fragment will be scaled together with upstream mv, do we want to add additional fragment breaker (exchange) for it? Specifically, will the join executor be scaled together with upstream?

Reply:
```
| mv1 | --no-shuffle-> | chain1 -> ... -> |           |
                                          | hash join |
| mv2 | --no-shuffle-> | chain2 -> ... -> |           |
```

vs

```
| mv1 | --no-shuffle-> | chain1 -> ... ->  hash join  |
                                          -------------
| mv2 | --no-shuffle-> | chain2 -> ...  ---^   
```

Reply:
err I think if `UpstreamHashShard` appears on any side, we should add an exchange. my point is that the current implementation in this PR could add unnecessary exchange

Reply:
Oh you mean e.g. dist key is superset of join key?

Reply:
```
dev=> explain select t.v1, t.v2, t.k from t join (
  select max(t.v1) as max_v1, k
  from t 
  group by k
) t1 on t.k = t1.k AND t.v1 = t1.max_v1;
                                 QUERY PLAN                                 
----------------------------------------------------------------------------
 BatchExchange { order: [], dist: Single }
   BatchHashJoin { type: Inner, predicate: t.k = t.k AND t.v1 = max(t.v1) }
     BatchExchange { order: [], dist: HashShard(t.k) }
       BatchScan { table: t, columns: [v1, v2, k] }
     BatchProject { exprs: [max(t.v1), t.k] }
       BatchHashAgg { group_key: [t.k], aggs: [max(t.v1)] }
         BatchExchange { order: [], dist: HashShard(t.k) }
           BatchProject { exprs: [t.k, t.v1] }
             BatchScan { table: t, columns: [v1, k] }
(9 rows)
```
the agg and join can be in same fregment

Reply:
tpch q18 is another good example https://github.com/singularity-data/risingwave/pull/3836